File: 125_valid_palindrome/solution.ts
// Use regex to determine if the char is alphanumeric
function isPalindrome1(s: string): boolean {
  const alphaNumericRegex = /^[A-Za-z0-9]$/;

  let head = 0;
  let tail = s.length - 1;

  while (head < tail) {
    const headCur = s[head];
    const tailCur = s[tail];

    if (!headCur.match(alphaNumericRegex)) {
      head++;
      continue;
    }
    if (!tailCur.match(alphaNumericRegex)) {
      tail--;
      continue;
    }

    console.log('head: ', headCur, 'tail: ', tailCur);

    if (headCur.toLowerCase() !== tailCur.toLowerCase()) return false;
    head++;
    tail--;
  }

  return true;
}

// Use self-made function to determine if the char is alphanumeric
function isAlphaNumeric(char: string) {
  const charCode = char.charCodeAt(0);

  // check uppercase
  if (charCode >= 'A'.charCodeAt(0) && charCode <= 'Z'.charCodeAt(0))
    return true;
  // check lowercase
  if (charCode >= 'a'.charCodeAt(0) && charCode <= 'z'.charCodeAt(0))
    return true;
  // check number
  if (charCode >= '0'.charCodeAt(0) && charCode <= '9'.charCodeAt(0))
    return true;

  return false;
}

function isPalindrome2(s: string): boolean {
  let head = 0;
  let tail = s.length - 1;

  while (head < tail) {
    const headCur = s[head];
    const tailCur = s[tail];

    if (!isAlphaNumeric(headCur)) {
      head++;
      continue;
    }
    if (!isAlphaNumeric(tailCur)) {
      tail--;
      continue;
    }

    console.log('head: ', headCur, 'tail: ', tailCur);

    if (headCur.toLowerCase() !== tailCur.toLowerCase()) return false;
    head++;
    tail--;
  }

  return true;
}

File: 125_valid_palindrome/solution.py
import re


class Solution:
    # Time: O(n), Space: O(n)
    def isPalindrome(self, s: str) -> bool:
        string = re.sub(r'[^a-zA-Z0-9]', '', s).lower()

        head = 0
        tail = len(string) - 1

        while head < tail:
            if string[head] != string[tail]:
                return False
            head += 1
            tail -= 1

        return True

    # Time: O(n), Space: O(1)
    def isPalindrome2(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l < r:
            while l < r and not self.isAlphaNumeric(s[l]):
                l += 1
            while l < r and not self.isAlphaNumeric(s[r]):
                r -= 1
            if s[l].lower() != s[r].lower():
                return False
            l += 1
            r -= 1
        return True

    def isAlphaNumeric(self, c: str) -> bool:
        return (
            ord("A") <= ord(c) <= ord("Z")
            or ord("a") <= ord(c) <= ord("z")
            or ord("0") <= ord(c) <= ord("9")
        )
